crates/executor/src/blockchain_tree/block_indices.rs

```diff
@@ -0,0 +1,284 @@
+//! Implementation of [`BlockIndices`] related to [`super::BlockchainTree`]
+
+use super::chain::{BlockChainId, Chain, ForkBlock};
+use reth_primitives::{BlockHash, BlockNumber, SealedBlockWithSenders};
+use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
+
+/// Internal indices of the blocks and chains.  This is main connection
+/// between blocks, chains and canonical chain.
+///
+/// It contains list of canonical block hashes, forks to childs blocks
+/// and block hash to chain id.
+#[derive(Clone)]
+pub struct BlockIndices {
+    /// Last finalized block.
+    last_finalized_block: BlockNumber,
+    /// For EVM's "BLOCKHASH" opcode we require last 256 block hashes. So we need to specify
+    /// at least `additional_canonical_block_hashes`+`max_reorg_depth`, for eth that would be
+    /// 256+64.
+    num_of_additional_canonical_block_hashes: u64,
+    /// Canonical chain. Contains N number (depends on `finalization_depth`) of blocks.
+    /// These blocks are found in fork_to_child but not inside `blocks_to_chain` or
+    /// `number_to_block` as those are chain specific indices.
+    canonical_chain: BTreeMap<BlockNumber, BlockHash>,
+    /// Index needed when discarding the chain, so we can remove connected chains from tree.
+    /// NOTE: It contains just a blocks that are forks as a key and not all blocks.
+    fork_to_child: HashMap<BlockHash, HashSet<BlockHash>>,
+    /// Block hashes and side chain they belong
+    blocks_to_chain: HashMap<BlockHash, BlockChainId>,
```

Comment:
No, `blocks_to_chain` contains only pending/side blocks. Hashes for the canonical chain are saved only in one place in `block_incides.canonical_chain` map

Reply:
could you amend the docs for the `contains_block_hash` function then? and perhaps rename it to `contains_fork_block_hash` or smth since it would return `false` for the canonical blocks

Reply:
That makes sense.